File: server/datasets/java_source-code/Multiple-G2/structural_33_avl_tree.java
class Node_8 {
    int k_8, h_8;
    Node_8 left, right;
    Node_8(int d) {
        k_8 = d;
        h_8 = 1;
    }
}
public class Tree_8 {
    public static void main(String[] args) {
        Tree_8 t = new Tree_8();
        Node_8 r = null;
        r = t.ins_8(r, 10);
    }
    int max(int a, int b) {
        return (a > b) ? a : b;
    }
    int h_8(Node_8 N) {
        if (N == null) return 0;
        return N.h_8;
    }
    int gb_8(Node_8 N) {
        if (N == null) return 0;
        return h_8(N.left) - h_8(N.right);
    }
    Node_8 upd_8(Node_8 n_8) {
        n_8.h_8 = 1 + max(h_8(n_8.left), h_8(n_8.right));
        return n_8;
    }
    Node_8 rr_8(Node_8 y) {
        Node_8 x = y.left;
        Node_8 T2 = x.right;
        x.right = y;
        y.left = T2;
        upd_8(y);
        upd_8(x);
        return x;
    }
    Node_8 rl_8(Node_8 x) {
        Node_8 y = x.right;
        Node_8 T2 = y.left;
        y.left = x;
        x.right = T2;
        upd_8(x);
        upd_8(y);
        return y;
    }
    Node_8 ins_8(Node_8 n_8, int k_8) {
        if (n_8 == null) return (new Node_8(k_8));
        if (k_8 < n_8.k_8) n_8.left = ins_8(n_8.left, k_8);
        else if (k_8 > n_8.k_8) n_8.right = ins_8(n_8.right, k_8);
        else return n_8;
        upd_8(n_8);
        int b_8 = gb_8(n_8);
        if (b_8 > 1 && k_8 < n_8.left.k_8) return rr_8(n_8);
        if (b_8 < -1 && k_8 > n_8.right.k_8) return rl_8(n_8);
        if (b_8 > 1 && k_8 > n_8.left.k_8) {
            n_8.left = rl_8(n_8.left);
            return rr_8(n_8);
        }
        if (b_8 < -1 && k_8 < n_8.right.k_8) {
            n_8.right = rr_8(n_8.right);
            return rl_8(n_8);
        }
        return n_8;
    }
}

File: server/datasets/controlled_benchmarks/python/multi_student_cohorts/cohort_g2/structural_33_avl_tree.py
class Node_8:
    def __init__(self, v_8):
        self.v_8 = v_8
        self.r_8 = None
        self.l_8 = None
        self.height = 1

def h_8(node):
    if not node:
        return 0
    return node.height

def b_8(node):
    if not node:
        return 0
    return h_8(node.l_8) - h_8(node.r_8)

def rr_8(y):
    x = y.l_8
    sub = x.r_8
    x.r_8 = y
    y.l_8 = sub
    y.height = 1 + max(h_8(y.l_8), h_8(y.r_8))
    x.height = 1 + max(h_8(x.l_8), h_8(x.r_8))
    return x

def rl_8(x):
    y = x.r_8
    sub = y.l_8
    y.l_8 = x
    x.r_8 = sub
    x.height = 1 + max(h_8(x.l_8), h_8(x.r_8))
    y.height = 1 + max(h_8(y.l_8), h_8(y.r_8))
    return y

def ins_8(rt_8, v_8):
    if not rt_8:
        return Node_8(v_8)
    if v_8 < rt_8.v_8:
        rt_8.l_8 = ins_8(rt_8.l_8, v_8)
    elif v_8 > rt_8.v_8:
        rt_8.r_8 = ins_8(rt_8.r_8, v_8)
    else:
        return rt_8

    rt_8.height = 1 + max(h_8(rt_8.l_8), h_8(rt_8.r_8))
    bf_8 = b_8(rt_8)

    if bf_8 > 1 and v_8 < rt_8.l_8.v_8:
        return rr_8(rt_8)
    if bf_8 < -1 and v_8 > rt_8.r_8.v_8:
        return rl_8(rt_8)
    if bf_8 > 1 and v_8 > rt_8.l_8.v_8:
        rt_8.l_8 = rl_8(rt_8.l_8)
        return rr_8(rt_8)
    if bf_8 < -1 and v_8 < rt_8.r_8.v_8:
        rt_8.r_8 = rr_8(rt_8.r_8)
        return rl_8(rt_8)

    return rt_8
